# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/2d/RecordedEvent.h
# Commit: 2b56c2b2837a
# Full Hash: 2b56c2b2837adfd7c49b599771cfea2d9e3d687b
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2023-11-22 21:50:37
# Regressor Bug: 1863914
# File Overlap Count: 2
# Description:
#   Bug 1863914: Use multiple shmem buffers for remote canvas recording. r=aosmond
#   
#   This replaces the use of a single large ring buffer.
#   The buffers are still processed in parallel and are recycled to reduce
#   allocation. Events that do not fit in the default sized buffer have a separate
# ==============================================================================

diff -r ed311995cd70 -r 2b56c2b2837a gfx/2d/RecordedEvent.h
--- a/gfx/2d/RecordedEvent.h	Wed Nov 22 14:06:43 2023 +0200
+++ b/gfx/2d/RecordedEvent.h	Wed Nov 22 11:52:08 2023 +0000
@@ -214,7 +214,7 @@
 };
 
 struct MemWriter {
-  explicit MemWriter(char* aPtr) : mPtr(aPtr) {}
+  constexpr explicit MemWriter(char* aPtr) : mPtr(aPtr) {}
   void write(const char* aData, size_t aSize) {
     memcpy(mPtr, aData, aSize);
     mPtr += aSize;
@@ -222,13 +222,30 @@
   char* mPtr;
 };
 
-// This is a simple interface for an EventRingBuffer, so we can use it in the
-// RecordedEvent reading and writing machinery.
-class EventRingBuffer {
+class ContiguousBuffer {
+ public:
+  ContiguousBuffer(char* aStart, size_t aSize)
+      : mWriter(aStart), mEnd(aStart + aSize) {}
+
+  constexpr MOZ_IMPLICIT ContiguousBuffer(std::nullptr_t) : mWriter(nullptr) {}
+
+  MemWriter& Writer() { return mWriter; }
+
+  size_t SizeRemaining() { return mWriter.mPtr ? mEnd - mWriter.mPtr : 0; }
+
+  bool IsValid() { return !!mWriter.mPtr; }
+
+ private:
+  MemWriter mWriter;
+  char* mEnd = nullptr;
+};
+
+// Allows a derived class to provide guaranteed contiguous buffer.
+class ContiguousBufferStream {
  public:
   /**
-   * Templated RecordEvent function so that when we have enough contiguous
-   * space we can record into the buffer quickly using MemWriter.
+   * Templated RecordEvent function so that we can record into the buffer
+   * quickly using MemWriter.
    *
    * @param aRecordedEvent the event to record
    */
@@ -237,56 +254,25 @@
     SizeCollector size;
     WriteElement(size, aRecordedEvent->GetType());
     aRecordedEvent->Record(size);
-    if (size.mTotalSize > mAvailable) {
-      WaitForAndRecalculateAvailableSpace();
+    auto& buffer = GetContiguousBuffer(size.mTotalSize);
+    if (!buffer.IsValid()) {
+      return;
     }
-    if (size.mTotalSize <= mAvailable) {
-      MemWriter writer(mBufPos);
-      WriteElement(writer, aRecordedEvent->GetType());
-      aRecordedEvent->Record(writer);
-      UpdateWriteTotalsBy(size.mTotalSize);
-    } else {
-      WriteElement(*this, aRecordedEvent->GetType());
-      aRecordedEvent->Record(*this);
-    }
-  }
 
-  /**
-   * Simple write function required by WriteElement.
-   *
-   * @param aData the data to be written to the buffer
-   * @param aSize the number of chars to write
-   */
-  virtual void write(const char* const aData, const size_t aSize) = 0;
+    MOZ_ASSERT(size.mTotalSize <= buffer.SizeRemaining());
 
-  /**
-   * Simple read function required by ReadElement.
-   *
-   * @param aOut the pointer to read into
-   * @param aSize the number of chars to read
-   */
-  virtual void read(char* const aOut, const size_t aSize) = 0;
-
-  virtual bool good() const = 0;
-
-  virtual void SetIsBad() = 0;
+    WriteElement(buffer.Writer(), aRecordedEvent->GetType());
+    aRecordedEvent->Record(buffer.Writer());
+    IncrementEventCount();
+  }
 
  protected:
   /**
-   * Wait until space is available for writing and then set mBufPos and
-   * mAvailable.
+   * Provide a contiguous buffer with at least aSize remaining.
    */
-  virtual bool WaitForAndRecalculateAvailableSpace() = 0;
+  virtual ContiguousBuffer& GetContiguousBuffer(size_t aSize) = 0;
 
-  /**
-   * Update write count, mBufPos and mAvailable.
-   *
-   * @param aCount number of bytes written
-   */
-  virtual void UpdateWriteTotalsBy(uint32_t aCount) = 0;
-
-  char* mBufPos = nullptr;
-  uint32_t mAvailable = 0;
+  virtual void IncrementEventCount() = 0;
 };
 
 struct MemStream {
@@ -430,7 +416,7 @@
 
   virtual void RecordToStream(std::ostream& aStream) const = 0;
   virtual void RecordToStream(EventStream& aStream) const = 0;
-  virtual void RecordToStream(EventRingBuffer& aStream) const = 0;
+  virtual void RecordToStream(ContiguousBufferStream& aStream) const = 0;
   virtual void RecordToStream(MemStream& aStream) const = 0;
 
   virtual void OutputSimpleEventInfo(std::stringstream& aStringStream) const {}
@@ -460,9 +446,6 @@
   static bool DoWithEventFromStream(
       EventStream& aStream, EventType aType,
       const std::function<bool(RecordedEvent*)>& aAction);
-  static bool DoWithEventFromStream(
-      EventRingBuffer& aStream, EventType aType,
-      const std::function<bool(RecordedEvent*)>& aAction);
 
   EventType GetType() const { return (EventType)mType; }
 
@@ -495,7 +478,7 @@
     WriteElement(aStream, this->mType);
     static_cast<const Derived*>(this)->Record(aStream);
   }
-  void RecordToStream(EventRingBuffer& aStream) const final {
+  void RecordToStream(ContiguousBufferStream& aStream) const final {
     aStream.RecordEvent(static_cast<const Derived*>(this));
   }
   void RecordToStream(MemStream& aStream) const override {